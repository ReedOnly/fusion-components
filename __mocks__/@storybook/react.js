import renderer from 'react-test-renderer';
import addons, { mockChannel } from '@storybook/addons';
import { withFusionContext } from '../../.storybook/withFusionContext';

addons.setChannel(mockChannel());

export const action = (actionName) => jest.fn();
const fusionContext = withFusionContext();
export const storiesOf = (groupName) => {
    const api = {
        decorators: [],
        parameters: {},

        add(storyName, story) {
            if (api.parameters.skipTest) {
                return;
            }

            describe(groupName, () => {
                it(storyName, () => {
                    // const decoratedStory = api.decorators.reduce((s, decorator) => {
                    //     try {
                    //         return decorator(s, { parameters: api.parameters });
                    //     } catch (e) {
                    //         return s;
                    //     }
                    // }, story);

                    const component = renderer.create(fusionContext(story));

                    // @TODO Enable custom names once released:
                    // > https://github.com/facebook/jest/pull/2094
                    expect(component.toJSON())
                        .toMatchSnapshot
                        // `${groupName}.${storyName}`
                        ();
                });
            });

            return api;
        },

        addDecorator() {
            return api;
        },

        addParameters(parameters) {
            api.parameters = { ...api.parameters, ...parameters };
            return api;
        },
    };

    return api;
};

export const addDecorator = () => {};
