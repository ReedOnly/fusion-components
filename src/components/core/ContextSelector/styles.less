.container {
    display: flex;
    align-items: center;
    padding: 0 calc(var(--grid-unit) * 2px);
}

.searchInput {
    min-width: calc(var(--grid-unit) * 50px);
    width: fit-content;
    padding: 0 calc(var(--grid-unit) * 1px);
    line-height: calc(var(--grid-unit) * 6px);
    height: calc(var(--grid-unit) * 6px);
    border: none;
    border-bottom: 1px solid transparent;
    outline: none;
    font-size: 16px;
    color: var(--color-primary-accent);
    background: none;
    box-sizing: border-box;
    margin-left: calc(var(--grid-unit) * 2px);

    &::placeholder {
        color: var(--color-primary-accent);
    }

    &.focus {
        border-bottom-color: var(--color-primary-accent);

        &::placeholder {
            color: var(--color-black-alt3);
        }
    }
}

.contextButton {
    padding: 0 calc(var(--grid-unit) * 1px);
    line-height: calc(var(--grid-unit) * 6px);
    height: calc(var(--grid-unit) * 6px);
    border: none;
    border-bottom: 1px solid transparent;
    outline: none;
    font-size: 16px;
    color: var(--color-primary-accent);
    background: none;
    text-align: center;
    box-sizing: border-box;
    margin-left: calc(var(--grid-unit) * 2px);
    cursor: pointer;
}

.popover {
    position: relative;
    top: 100%;
    margin-top: calc(var(--grid-unit) * 1px);
    pointer-events: all;
}

.noResultsMessage {
    background: var(--color-white);
    border-radius: 4px;
    padding: calc(var(--grid-unit) * 2px);
    max-width: calc(var(--grid-unit) * 70px)
}