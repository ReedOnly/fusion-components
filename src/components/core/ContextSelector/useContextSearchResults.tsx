import * as React from 'react';
import {
    useContextQuery,
    ContextTypes,
    useContextManager,
    useCurrentContext,
    Context,
} from '@equinor/fusion';
import Menu, { MenuSection } from '../../general/Menu';
import useRelativePortal from '../../../hooks/useRelativePortal';
import useClickOutsideOverlayPortal from '../../../hooks/useClickOutsideOverlayPortal';
import useElevationClassName from '../../../hooks/useElevationClassName';
import classNames from 'classnames';
import styles from "./styles.less";

export default (
    inputRef: HTMLElement | null,
    containerRef: React.RefObject<HTMLElement>,
    types: ContextTypes[],
    relatedContexts: Context[],
    setCurrentContextAsync: (context: Context | undefined) => Promise<void>,
    queryText: string,
    shouldShow: boolean,
    onClose: () => void
): boolean => {
    const currentContext = useCurrentContext();
    const contextManager = useContextManager();
    const popoverRef = React.useRef(null);
    const [queryError, isQuerying, contexts, queryContexts] = useContextQuery(...types);
    const [history, setHistory] = React.useState<Context[]>([]);

    React.useEffect(() => {
        queryContexts(queryText);
    }, [queryText]);

    React.useEffect(() => {
        contextManager.getAsync('history').then(history => {
            if (!history) {
                return;
            }

            setHistory(history.filter(c => types.find(type => c.type.id === type)));
        });
    }, [currentContext]);

    const mapContextToMenuItem = (context: Context) => ({
        key: context.id,
        title: context.title,
        aside: context.type.id, // TODO: Should get a displayName from the backend in the future
        isSelected: currentContext ? context.id === currentContext.id : false,
    });

    const sections: MenuSection[] = [];

    // Add context search result to results menu sections
    if (contexts.length && queryText.length) {
        sections.push({
            key: 'search results',
            title: relatedContexts.length > 0 || history.length > 0 ? 'Search results' : '',
            items: contexts.map(mapContextToMenuItem),
        });
    }

    // Add related contexts to results menu sections
    if (relatedContexts.length) {
        sections.push({
            key: 'exchange context',
            title: currentContext
                ? `Exchange ${currentContext.title} with one of the following:`
                : 'Select context',
            items: relatedContexts.map(mapContextToMenuItem),
        });
    }

    // Add context history to results menu sections if nothing else is displayed
    if (history.length && shouldShow && (queryText.length === 0 || isQuerying) && sections.length === 0) {
        sections.push({
            key: 'history',
            title: 'Recent searches',
            items: history.map(mapContextToMenuItem),
        });
    }

    const onSelectContext = React.useCallback(
        item => {
            const context = [...contexts, ...relatedContexts, ...history].find(
                context => context.id === item.key
            );
            setCurrentContextAsync(context);
        },
        [contexts, relatedContexts, history, contextManager]
    );

    const [showNoResultsMessage, setShowNoResultsMessage] = React.useState(false);
    React.useEffect(() => {
        setShowNoResultsMessage(!isQuerying && queryText.length > 2);
    }, [isQuerying, queryText]);

    const getPopoverContent = () => {
        if (sections.length) {
            return (
                <Menu
                    sections={sections}
                    onClick={onSelectContext}
                    keyboardNavigationRef={inputRef}
                />
            );
        }

        if(showNoResultsMessage) {
            const noResultsClassNames = classNames(styles.noResultsMessage, useElevationClassName(2));
            return (
                <div className={noResultsClassNames}>
                    <h5>No results for <strong>{queryText}</strong></h5>
                </div>
            );
        }
    };

    const popoverContent = (
        <div
            ref={popoverRef}
            className={styles.popover}
        >
            {getPopoverContent()}
        </div>
    );

    useRelativePortal(
        popoverContent,
        containerRef,
        shouldShow || (queryText.length > 0 && contexts.length > 0) || relatedContexts.length > 0
    );

    useClickOutsideOverlayPortal(onClose, popoverRef.current, containerRef.current);

    return isQuerying;
};
