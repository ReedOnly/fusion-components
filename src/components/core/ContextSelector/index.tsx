import * as React from 'react';
import {
    ContextTypes,
    useContextManager,
    useCurrentContext,
    Context,
    useComponentDisplayType,
    ComponentDisplayType,
} from '@equinor/fusion';
import useContextSearchResults from './useContextSearchResults';
import classNames from 'classnames';
import styles from './styles.less';
import { SearchIcon } from '../../icons/components/action';
import { Spinner } from 'index';

type ContextSelectorProps = {
    types: ContextTypes[];
};

const ContextSelector: React.FC<ContextSelectorProps> = ({ types }: ContextSelectorProps) => {
    const [shouldShowResultsDropdown, setShouldShowResultsDropdown] = React.useState(false);
    const [queryText, setQueryText] = React.useState('');
    const [relatedContexts, setRelatedContexts] = React.useState<Context[]>([]);
    const [isFetchingRelatedContexts, setIsFetchingRelatedContexts] = React.useState(false);
    const [inputRef, setInputRef] = React.useState<HTMLElement | null>(null);
    const contextManager = useContextManager();
    const currentContext = useCurrentContext();
    const currentContextForType = useCurrentContext(...types);

    const exchangeCurrentContext = async () => {
        setIsFetchingRelatedContexts(true);

        try {
            const possibleContexts = await contextManager.exchangeCurrentContextAsync(...types);
            const linkedContext = currentContext
                ? await contextManager.getLinkedContextAsync(currentContext)
                : null;

            setIsFetchingRelatedContexts(false);

            if (possibleContexts.length === 1) {
                return contextManager.setCurrentContextAsync(possibleContexts[0]);
            }

            const linkedContextIsRelevant =
                linkedContext &&
                types.find(type => linkedContext.type.id === type) &&
                possibleContexts.find(c => c.id === linkedContext.id);
            if (linkedContext && linkedContextIsRelevant) {
                return contextManager.setCurrentContextAsync(linkedContext);
            }

            if (!possibleContexts.length) {
                return;
            }

            // Try to resolve a related context from the history
            const history = await contextManager.getAsync('history');
            if (history !== null && history.length > 0) {
                const fromHistory = history.filter(
                    c => c && possibleContexts.find(pc => pc.id === c.id)
                );

                if (fromHistory.length === 1) {
                    return contextManager.setCurrentContextAsync(fromHistory[0]);
                }
            }

            setRelatedContexts(possibleContexts);
        } catch {
            setIsFetchingRelatedContexts(false);
        }

        if (inputRef) {
            inputRef.focus();
        }
    };

    React.useEffect(() => {
        if (currentContext && !types.find(type => currentContext.type.id === type)) {
            exchangeCurrentContext();
        }
    }, [currentContext, types]);

    const setCurrentContextAsync = async (context?: Context) => {
        if (context) {
            await contextManager.setCurrentContextAsync(context);
            setQueryText('');
            setRelatedContexts([]);
            setShouldShowResultsDropdown(false);
        }
    };

    const onSearchChange = React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
        setQueryText(e.target.value);
    }, []);

    const onOpen = React.useCallback(() => setShouldShowResultsDropdown(true), []);
    const onClose = React.useCallback(() => {
        setShouldShowResultsDropdown(false);
        setQueryText('');
    }, []);

    const onKeyUp = React.useCallback((e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.keyCode === 27) {
            // ESC
            onClose();
        }
    }, []);

    React.useEffect(() => {
        if (shouldShowResultsDropdown && inputRef) {
            inputRef.focus();
        }
    }, [shouldShowResultsDropdown, inputRef]);

    const containerRef = React.useRef<HTMLDivElement>(null);

    const isQuerying = useContextSearchResults(
        inputRef,
        containerRef,
        types,
        relatedContexts,
        setCurrentContextAsync,
        queryText,
        shouldShowResultsDropdown,
        onClose
    );

    const inputClassNames = classNames(styles.searchInput, {
        [styles.focus]: shouldShowResultsDropdown,
    });

    const getButtonContent = () => {
        if (currentContextForType) {
            return currentContextForType.title;
        }

        if (isFetchingRelatedContexts) {
            return (
                <span>
                    Resolving {currentContext ? currentContext.title : null} <Spinner inline />
                </span>
            );
        }

        if (currentContext && !isFetchingRelatedContexts) {
            return `Unable to resolve ${currentContext.title}. Please select context`;
        }

        return 'Select context';
    };

    const componentDisplayType = useComponentDisplayType();
    const containerClassNames = classNames(styles.container, {
        [styles.comfortable]: componentDisplayType === ComponentDisplayType.Comfortable,
        [styles.compact]: componentDisplayType === ComponentDisplayType.Compact,
    });

    return (
        <div ref={containerRef} onClick={onOpen} className={containerClassNames}>
            <SearchIcon color="#DADADA" />
            {shouldShowResultsDropdown ? (
                <>
                    <input
                        ref={setInputRef}
                        type="text"
                        value={queryText}
                        onChange={onSearchChange}
                        onKeyUp={onKeyUp}
                        placeholder={'Search'}
                        className={inputClassNames}
                    />
                    {isQuerying && <Spinner inline />}
                </>
            ) : (
                <button className={styles.contextButton}>{getButtonContent()}</button>
            )}
        </div>
    );
};

export default ContextSelector;
