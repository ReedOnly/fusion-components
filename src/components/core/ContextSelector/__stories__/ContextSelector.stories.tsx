import * as React from "react";
import { storiesOf } from "@storybook/react";
import { ContextTypes } from "@equinor/fusion";
import ContextSelector from "../index";

const PDPContextSelectorStory = () => <ContextSelector type={ContextTypes.PDP} />;
const OrgChartContextSelectorStory = () => <ContextSelector type={ContextTypes.OrgChart} />;

storiesOf("Core components/Context Selector", module).add("PDP", () => (
    <PDPContextSelectorStory />
));

storiesOf("Core components/Context Selector", module).add("OrgChart", () => (
    <OrgChartContextSelectorStory />
));
