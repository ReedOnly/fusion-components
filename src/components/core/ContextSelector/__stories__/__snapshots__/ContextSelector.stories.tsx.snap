// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core|Context Selector PDP 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
     
    <div
      className="container comfortable"
    />
    <div
      className="container"
    >
      <h1
        style={
          Object {
            "margin": 16,
            "textAlign": "center",
          }
        }
      >
        PDP
      </h1>
    </div>
     
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`Core|Context Selector PDP Nullable And Placeholder 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
     
    <div
      className="container comfortable"
    />
    <div
      className="container"
    >
      <h1
        style={
          Object {
            "margin": 16,
            "textAlign": "center",
          }
        }
      >
        PDP
      </h1>
    </div>
     
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
